File: gfx2Imp/java/PixmapImp.java
//
// Copyright (c) 2011, chunquedong
// Licensed under the Academic Free License version 3.0
//
// History:
//   2011-08-14  Jed Young  Creation
//

package fan.gfx2Imp;

import fan.sys.*;
import fan.gfx.*;
import fan.gfx2.*;

import java.io.*;

import org.eclipse.swt.*;
import org.eclipse.swt.graphics.*;
import fanx.interop.Interop;

public class PixmapImp implements Pixmap
{
  private boolean imageChanged = false;
  private org.eclipse.swt.graphics.Image image;
  private org.eclipse.swt.graphics.ImageData imageData;
  public org.eclipse.swt.graphics.Image getImage()
  {
    if (!imageChanged) return image;

    //create new image
    image.dispose();
    image = new org.eclipse.swt.graphics.Image(FwtEnv2Peer.getDisplay(), imageData);
    imageChanged = false;
    return image;
  }

  public PixmapImp(org.eclipse.swt.graphics.Image m)
  {
    image = m;
    imageData = m.getImageData();
  }

  public Size size()
  {
    //Rectangle r = image.getBounds();
    return Size.make(imageData.width, imageData.height);
  }

  public fan.gfx.Color getPixel(long x, long y)
  {
    ImageData data= imageData;
    PaletteData palette = data.palette;
    RGB rgb = palette.getRGB(data.getPixel((int)x, (int)y));
    int alpha = data.getAlpha((int)x, (int)y);
    return fan.gfx.Color.makeArgb(alpha, rgb.red, rgb.green, rgb.blue);
  }
  public void setPixel(long x, long y, fan.gfx.Color value)
  {
    ImageData data= imageData;
    PaletteData palette = data.palette;
    RGB rgb = new RGB((int)value.r(), (int)value.g(), (int)value.b());
    int alpha = (int)value.a();
    int pixel = palette.getPixel(rgb);
    data.setPixel((int)x, (int)y, pixel);
    //System.out.println(pixel);
    data.setAlpha((int)x, (int)y, alpha);
    imageChanged = true;
  }

  public fan.gfx.Image toImage() { throw UnsupportedErr.make(); }

  /**
   * save image to outSteam
   */
  public void save(OutStream out, MimeType format)
  {
    ImageLoader loader = new ImageLoader();
    loader.data = new ImageData[] { imageData };
    OutputStream jout = Interop.toJava(out);

    int swtFormat = SWT.IMAGE_PNG;
    String subType = format.subType();
    if (subType.equals("png")) swtFormat = SWT.IMAGE_PNG;
    else if (subType.equals("gif")) swtFormat = SWT.IMAGE_GIF;
    else if (subType.equals("jpeg")) swtFormat = SWT.IMAGE_JPEG;
    else if (subType.equals("jpg")) swtFormat = SWT.IMAGE_JPEG;
    else if (subType.equals("bmp")) swtFormat = SWT.IMAGE_BMP;
    else if (subType.equals("tiff")) swtFormat = SWT.IMAGE_TIFF;
    else if (subType.equals("ico")) swtFormat = SWT.IMAGE_ICO;
    else throw UnsupportedErr.make("unsupported image type: "+subType);

    loader.save(jout, swtFormat);
  }
  public void save(OutStream out)
  {
    save(out, MimeType.forExt("png"));
  }
  public boolean isLoaded() { return true; }

  /**
   * get graphics context from image
   */
  public Graphics2 graphics()
  {
    int w = (int)size().w;
    int h = (int)size().h;
    return new FwtGraphics2(new GC(getImage()), 0, 0, w, h);
  }
}

File: gfx2Imp/js/PixmapImp.js
//
// Copyright (c) 2011, chunquedong
// Licensed under the Academic Free License version 3.0
//
// History:
//   2011-08-20  Jed Young  Creation
//

fan.gfx2Imp.PixmapImp = fan.sys.Obj.$extend(fan.sys.Obj);
fan.gfx2Imp.PixmapImp.prototype.$ctor = function() {}

fan.gfx2Imp.PixmapImp.prototype.$typeof = function()
{
  return fan.gfx2.Pixmap.$type;
}

fan.gfx2Imp.PixmapImp.prototype.m_imageData = null;
fan.gfx2Imp.PixmapImp.prototype.m_image = null;
fan.gfx2Imp.PixmapImp.prototype.m_isImageData = false;
fan.gfx2Imp.PixmapImp.prototype.m_isLoaded = false;
fan.gfx2Imp.PixmapImp.prototype.m_uri = null;
fan.gfx2Imp.PixmapImp.prototype.m_imageChanged = false;
fan.gfx2Imp.PixmapImp.prototype.m_painted = false;

fan.gfx2Imp.PixmapImp.prototype.m_size = null;
fan.gfx2Imp.PixmapImp.prototype.size = function() { return this.m_size; }

fan.gfx2Imp.PixmapImp.prototype.getImage = function(widget)
{
  if (this.m_image && !this.m_imageChanged)
  {
    this.m_imageChanged = false;
    return this.m_image;
  }

  if (!this.m_isImageData && !this.m_isLoaded)
  {
    fan.fwt.FwtEnvPeer.loadImage(this, widget);
    return;
  }

  var canvas = this.getCanvas();
  if(!this.m_painted)
  {
    this.m_cx = canvas.getContext("2d");
    this.m_cx.putImageData(this.m_imageData, 0, 0);
    this.m_painted = false;
  }
  this.m_uri = fan.sys.Uri.fromStr(canvas.toDataURL());

  this.m_image = fan.fwt.FwtEnvPeer.loadImage(this, widget);
  return this.m_image;
}

fan.gfx2Imp.PixmapImp.prototype.getImageData = function()
{
  if (this.m_imageData) return this.m_imageData;

  var canvas = this.getCanvas();
  this.m_cx = canvas.getContext("2d");
  if (this.m_image) this.m_cx.drawImage(this.m_image, 0, 0);
  this.m_imageData = this.m_cx.getImageData(0, 0, this.m_size.m_w, this.m_size.m_h);
  return this.m_imageData;
}

fan.gfx2Imp.PixmapImp.make = function(size)
{
  var p = new fan.gfx2Imp.PixmapImp();
  p.m_size = size;
  p.m_canvas = document.createElement("canvas");
  p.m_canvas.width = size.m_w;
  p.m_canvas.height = size.m_h;
  p.m_cx = p.m_canvas.getContext("2d");
  p.m_imageData = p.m_cx.getImageData(0, 0, size.m_w, size.m_h);
  p.m_isImageData = true;
  p.m_isLoaded = true;
  return p;
}

fan.gfx2Imp.PixmapImp.prototype.getPixel = function(x, y)
{
  var index = (y * this.getImageData().width + x)*4;
  var r = this.getImageData().data[index];
  var g = this.getImageData().data[index +1];
  var b = this.getImageData().data[index +2];
  var a = this.getImageData().data[index +3];
  return fan.gfx.Color.makeArgb(a, r, g, b);
}

fan.gfx2Imp.PixmapImp.prototype.setPixel = function(x, y, value)
{
  var index = (y * this.getImageData().width + x)*4;
  this.getImageData().data[index] = value.r();
  this.getImageData().data[index+1] = value.g();
  this.getImageData().data[index+2] = value.b();
  this.getImageData().data[index+3] = value.a();

  this.m_imageChanged = true;
}

fan.gfx2Imp.PixmapImp.prototype.toImage = function()
{
  throw fan.sys.UnsupportedErr.make();
}

fan.gfx2Imp.PixmapImp.prototype.getCanvas = function()
{
  if (!this.m_canvas)
  {
    this.m_canvas = document.createElement("canvas");
    this.m_canvas.width = this.m_size.m_w;
    this.m_canvas.height = this.m_size.m_h;
  }
  return this.m_canvas;
}

fan.gfx2Imp.PixmapImp.prototype.graphics = function()
{
  var canvas = this.getCanvas();
  var g = new fan.gfx2Imp.Graphics2();
  var imageData = this.m_imageData;
  var image = this.m_image;
  g.paint(canvas, fan.gfx.Rect.make(0, 0, this.m_size.m_w, this.m_size.m_h), function()
  {
    if (imageData)
     g.cx.putImageData(imageData, 0, 0);
    else if(image)
     g.cx.drawImage(image, 0, 0);
  });
  this.m_painted = true;
  this.m_imageChanged = true;
  return g;
}

fan.gfx2Imp.PixmapImp.prototype.flush = function()
{
  this.m_uri = fan.sys.Uri.fromStr(canvas.toDataURL());
  this.m_image = fan.fwt.FwtEnvPeer.loadImage(this, widget);
}

fan.gfx2Imp.PixmapImp.prototype.save = function(out, format)
{
  //TODO
}

fan.gfx2Imp.PixmapImp.prototype.isLoaded = function() { return this.m_isLoaded; }

